File: src/com/mojang/ld22/save/ItemIO.java
package com.mojang.ld22.save;

import com.google.gwt.json.client.JSONBoolean;
import com.google.gwt.json.client.JSONNumber;
import com.google.gwt.json.client.JSONObject;
import com.google.gwt.json.client.JSONString;
import com.mojang.ld22.entity.Furniture;
import com.mojang.ld22.item.FurnitureItem;
import com.mojang.ld22.item.Item;
import com.mojang.ld22.item.PowerGloveItem;
import com.mojang.ld22.item.ResourceItem;
import com.mojang.ld22.item.ToolItem;
import com.mojang.ld22.item.ToolType;
import com.mojang.ld22.item.resource.Resource;

/**
 * Serializes {@link Item}s by {@code kind}, round-tripping through stable
 * name-based identifiers (Resource.name / ToolType.name / Furniture.name).
 */
public final class ItemIO {

    private ItemIO() {
    }

    public static JSONObject write(Item item) {
        JSONObject o = new JSONObject();
        if (item instanceof ResourceItem) {
            ResourceItem ri = (ResourceItem) item;
            o.put("kind", new JSONString("resource"));
            o.put("resourceName", new JSONString(ri.resource.name));
            o.put("count", new JSONNumber(ri.count));
        } else if (item instanceof ToolItem) {
            ToolItem ti = (ToolItem) item;
            o.put("kind", new JSONString("tool"));
            o.put("toolName", new JSONString(ti.type.name));
            o.put("level", new JSONNumber(ti.level));
        } else if (item instanceof FurnitureItem) {
            FurnitureItem fi = (FurnitureItem) item;
            o.put("kind", new JSONString("furniture"));
            o.put("furnitureName", new JSONString(fi.furniture.name));
            o.put("placed", JSONBoolean.getInstance(fi.placed));
        } else if (item instanceof PowerGloveItem) {
            o.put("kind", new JSONString("powerglove"));
        }
        return o;
    }

    public static Item read(JSONObject o) {
        if (o == null) return null;
        String kind = o.get("kind").isString().stringValue();
        if ("resource".equals(kind)) {
            Resource r = Resource.getByName(o.get("resourceName").isString().stringValue());
            if (r == null) return null;
            ResourceItem ri = new ResourceItem(r);
            ri.count = (int) o.get("count").isNumber().doubleValue();
            return ri;
        } else if ("tool".equals(kind)) {
            ToolType t = ToolType.getByName(o.get("toolName").isString().stringValue());
            if (t == null) return null;
            int lvl = (int) o.get("level").isNumber().doubleValue();
            return new ToolItem(t, lvl);
        } else if ("furniture".equals(kind)) {
            Furniture f = Furniture.createByName(o.get("furnitureName").isString().stringValue());
            if (f == null) return null;
            FurnitureItem fi = new FurnitureItem(f);
            fi.placed = o.get("placed").isBoolean().booleanValue();
            return fi;
        } else if ("powerglove".equals(kind)) {
            return new PowerGloveItem();
        }
        return null;
    }
}

File: ts-port/src/game/save/ItemIO.ts
/**
 * ItemIO — serialises the four item variants to/from plain JSON objects.
 *
 * Dispatch key is `kind`:
 *   'resource'    -> ResourceItem   { resource: Resource.name, count }
 *   'tool'        -> ToolItem       { type: ToolType.name, level }
 *   'furniture'   -> FurnitureItem  { furniture: Furniture.name, placed }
 *   'powerglove'  -> PowerGloveItem (no payload)
 *
 * Resources/Tools/Furniture are resolved back through their existing
 * `getByName` registries, so the persisted form is name-based and stable
 * across process runs. Unknown payloads round-trip to null and are dropped by
 * the caller (defensive against forward-incompatible saves).
 */
import type { Item } from '../item/Item';
import { ResourceItem } from '../item/ResourceItem';
import { ToolItem } from '../item/ToolItem';
import { FurnitureItem } from '../item/FurnitureItem';
import { PowerGloveItem } from '../item/PowerGloveItem';
import { Resource } from '../item/resource/Resource';
import { ToolType } from '../item/ToolType';
import { Furniture } from '../entity/Furniture';

export class ItemIO {
  /** Serialise an item, or null if it has no persistent representation. */
  public static write(item: Item): Record<string, unknown> | null {
    if (item instanceof ResourceItem) {
      return { kind: 'resource', resource: item.resource.name, count: item.count };
    }
    if (item instanceof ToolItem) {
      return { kind: 'tool', type: item.type.name, level: item.level };
    }
    if (item instanceof FurnitureItem) {
      return { kind: 'furniture', furniture: item.furniture.name, placed: item.placed };
    }
    if (item instanceof PowerGloveItem) {
      return { kind: 'powerglove' };
    }
    return null;
  }

  /** Reconstruct an item, or null if the payload is unrecognised. */
  public static read(o: unknown): Item | null {
    if (!o || typeof o !== 'object') return null;
    const obj = o as Record<string, unknown>;
    const kind = obj.kind;
    switch (kind) {
      case 'resource': {
        const r = Resource.getByName(obj.resource as string);
        if (!r) return null;
        return new ResourceItem(r, (obj.count as number) ?? 1);
      }
      case 'tool': {
        const t = ToolType.getByName(obj.type as string);
        if (!t) return null;
        return new ToolItem(t, (obj.level as number) ?? 0);
      }
      case 'furniture': {
        const f = Furniture.createByName(obj.furniture as string);
        if (!f) return null;
        const it = new FurnitureItem(f);
        it.placed = obj.placed === true;
        return it;
      }
      case 'powerglove':
        return new PowerGloveItem();
      default:
        return null;
    }
  }
}
